cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialProfileStore.java

```diff
@@ -0,0 +1,172 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.credential.store;
+
+import com.google.gson.Gson;
+import io.cdap.cdap.api.dataset.lib.CloseableIterator;
+import io.cdap.cdap.common.AlreadyExistsException;
+import io.cdap.cdap.common.NotFoundException;
+import io.cdap.cdap.proto.credential.CredentialProfile;
+import io.cdap.cdap.proto.id.CredentialProfileId;
+import io.cdap.cdap.spi.data.StructuredRow;
+import io.cdap.cdap.spi.data.StructuredTable;
+import io.cdap.cdap.spi.data.StructuredTableContext;
+import io.cdap.cdap.spi.data.table.field.Field;
+import io.cdap.cdap.spi.data.table.field.Fields;
+import io.cdap.cdap.spi.data.table.field.Range;
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import io.cdap.cdap.spi.data.transaction.TransactionRunners;
+import io.cdap.cdap.store.StoreDefinition.CredentialProvisionerStore;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Optional;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+import javax.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Storage for credential identities.
+ */
+public class CredentialProfileStore {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CredentialProfileStore.class);
+  private static final Gson GSON = new Gson();
+
+  private final TransactionRunner transactionRunner;
+
+  @Inject
+  CredentialProfileStore(TransactionRunner transactionRunner) {
+    this.transactionRunner = transactionRunner;
+  }
+
+  public Collection<CredentialProfileId> list(String namespace) throws IOException {
+    return TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable table = context.getTable(CredentialProvisionerStore.CREDENTIAL_PROFILES);
+      Field<?> namespaceIndex = Fields.stringField(CredentialProvisionerStore.NAMESPACE_FIELD,
+          namespace);
+      try (CloseableIterator<StructuredRow> iterator = table.scan(Range.singleton(
+          Collections.singleton(namespaceIndex)), Integer.MAX_VALUE)) {
+        return profilesFromRowIterator(iterator);
+      }
+    }, IOException.class);
+  }
+
+  public Optional<CredentialProfile> get(CredentialProfileId id) throws IOException {
+    Utils.validateResourceName(id.getName());
+    return TransactionRunners.run(transactionRunner, context -> {
+      Optional<StructuredRow> row = fetchRow(context, id);
+      if (!row.isPresent()) {
```

Comment:
similar comment in this class about `map`

Reply:
Done